gobblin-api/src/main/java/org/apache/gobblin/stream/ControlMessageInjector.java

```diff
@@ -0,0 +1,154 @@
+ * the input record
+ * @param <SI>
+ * @param <DI>
+ */
+public abstract class ControlMessageInjector<SI, DI> implements Closeable,
+        RecordStreamProcessor<SI, SI, DI, DI> {
+
+  @Setter(AccessLevel.PROTECTED)
+  @Getter(AccessLevel.PROTECTED)
+  private GlobalMetadata<SI> inputGlobalMetadata;
+
+  /**
+   * Initialize this {@link ControlMessageInjector}.
+   *
+   * @param workUnitState a {@link WorkUnitState} object carrying configuration properties
+   * @return an initialized {@link ControlMessageInjector} instance
+   */
+  protected ControlMessageInjector<SI, DI> init(WorkUnitState workUnitState) {
+    return this;
+  }
+
+  @Override
+  public void close() throws IOException {
+  }
+
+  /**
+   * Set the global metadata of the input messages
+   * @param inputGlobalMetadata the global metadata for input messages
+   * @param workUnitState
+   */
+  protected void setInputGlobalMetadata(GlobalMetadata<SI> inputGlobalMetadata, WorkUnitState workUnitState) {
+    this.inputGlobalMetadata = inputGlobalMetadata;
+  }
+
+  /**
+   * Inject {@link ControlMessage}s before the record
+   * @param inputRecordEnvelope
+   * @param workUnitState
+   * @return The {@link ControlMessage}s to inject before the record
+   */
+  protected abstract Iterable<ControlMessage<DI>> injectControlMessagesBefore(RecordEnvelope<DI> inputRecordEnvelope,
+      WorkUnitState workUnitState);
+
+  /**
+   * Inject {@link ControlMessage}s after the record
+   * @param inputRecordEnvelope
+   * @param workUnitState
+   * @return The {@link ControlMessage}s to inject after the record
+   */
+  protected abstract Iterable<ControlMessage<DI>> injectControlMessagesAfter(RecordEnvelope<DI> inputRecordEnvelope,
+      WorkUnitState workUnitState);
+
+  /**
+   * Apply injections to the input {@link RecordStreamWithMetadata}.
+   * {@link ControlMessage}s may be injected before, after, or around the input record.
+   * A {@link MetadataUpdateControlMessage} will update the current input {@link GlobalMetadata} and pass the
+   * updated input {@link GlobalMetadata} to the next processor to propagate the metadata update down the pipeline.
+   */
+  @Override
+  public RecordStreamWithMetadata<DI, SI> processStream(RecordStreamWithMetadata<DI, SI> inputStream,
+      WorkUnitState workUnitState) throws StreamProcessingException {
+    init(workUnitState);
+
+    setInputGlobalMetadata(inputStream.getGlobalMetadata(), workUnitState);
+
+    Flowable<StreamEntity<DI>> outputStream =
+        inputStream.getRecordStream()
+            .flatMap(in -> {
+              if (in instanceof ControlMessage) {
+                ControlMessage out = (ControlMessage) in;
+                if (in instanceof MetadataUpdateControlMessage) {
+                  setInputGlobalMetadata(((MetadataUpdateControlMessage) in).getGlobalMetadata(),
+                      workUnitState);
+                  out = new MetadataUpdateControlMessage<SI, DI>(this.inputGlobalMetadata);
```

Comment:
Removed inputGlobalMetadata from this class so that there is no confusion over the propagation of metadata. A `ControlMessageInjector` can store a modified copy of the metadata, but it won't be able to pass that change along to the next construct. Only converters should change schemas.